src/swarm-lang/Swarm/Language/LSP/Definition.hs

```diff
@@ -0,0 +1,117 @@
+{-# LANGUAGE OverloadedStrings #-}
+
+-- LSP support for finding and jumping to definitions
+-- SPDX-License-Identifier: BSD-3-Clause
+module Swarm.Language.LSP.Definition (
+  findDefinition,
+  DefinitionResult (..),
+) where
+
+import Control.Monad.IO.Class (MonadIO (liftIO))
+import Data.Functor (($>))
+import Data.List.NonEmpty qualified as NE
+import Data.Maybe (mapMaybe, maybeToList)
+import Data.Text (Text)
+import Data.Text qualified as T
+import Data.Text.IO qualified as Text
+import Data.Text.Utf16.Rope.Mixed qualified as R
+import Debug.Trace (traceShow)
+import GHC.IO (unsafePerformIO)
+import Language.LSP.Protocol.Types qualified as J
+import Language.LSP.Protocol.Types qualified as LSP
+import Language.LSP.VFS (VirtualFile (VirtualFile), virtualFileText)
+import Swarm.Language.LSP.Position qualified as P
+import Swarm.Language.Parser (readTerm')
+import Swarm.Language.Parser.Core (ParserError, defaultParserConfig)
+import Swarm.Language.Pipeline (processParsedTerm)
+import Swarm.Language.Syntax (Located (..), SrcLoc, Syntax, Syntax' (Syntax'), Term' (..), Var)
+import Swarm.Language.TDVar (tdVarName)
+import Swarm.Language.Typecheck (ContextualTypeErr)
+import System.IO (stderr)
+
+data DefinitionResult = TError ContextualTypeErr | PError ParserError | Unsupported | NotFound | Found [J.Range]
+
+findDefinition ::
+  J.NormalizedUri ->
+  J.Position ->
+  VirtualFile ->
+  DefinitionResult
+findDefinition _ p vf@(VirtualFile _ _ myRope) =
```

Comment:
I'm not super happy with this function. It seems to be doing a lot and the multiple wheres don't seem that right to me. But it works so maybe ok?